### DISCUSSION TITLE
Is Kratos sufficient enough for an internal auth service?

### DISCUSSION THREAD

**Discussion by elirud** [ory/kratos#3523] [Q&A] [answered]
So I've started to develop a simple auth service for internal use, where the most important thing is the usage of the same credentials between different applications. Having "true" SSO where you stay logged in between applications is not as crucial, but would be nice on the frontend apps of course (which would be on the same top level domain). There are no plans for any third party access, and if so, Hydra is possible to add on top of the service if need be(?). This would only be for our own frontend applications / various python scripts / CLI tools to use for authorization when calling our backend APIs.

[This blogpost](https://www.ory.sh/oauth2-openid-connect-do-you-need-use-cases-examples/) really made sense to me, and it feels like the more I look into Oauth / OIDC / SAML, it feels like it would be overkill to set it up. I have looked at Keycloak as well, but I like what I'm seeing and reading about the Ory ecosystem and the modularity.

I took a look at [this example](https://www.ory.sh/securing-flask-application-using-kratos-and-keto/) using Kratos and Keto for securing a flask application. The only thing that's got me hesitating is if it's fine to set this up in a "SSO kind of way", because people also mention that if you want SSO for usage with multiple applications, you should use Oauth, so Hydra I would assume. But setting up Hydra as well together with Kratos and Keto, that would probably have me looking again at Keycloak since I feel like it would be simpler. But maybe that is only when working towards "true" SSO?

Anyways, sorry for the rambling, my question is basically if Kratos + Keto is a sufficient and "proper" enough solution for what I'm trying to do here, or if I should turn back and get something else up and running.

**Comment by vinckr** [comment, marked as answer]
Hello @elirud 

Short answer: Yes Ory Kratos is probably enough to build an internal auth service!

> SSO kind of way

What exactly do you mean by that? Setting up 1-click login via OIDC or webauthn is easy with Kratos. What you need Hydra for is if you want to become an SSO provider yourself; so for example offer "Login with $YOURCORP".

In general I would recommend to check out [Ory Network](https://console.ory.sh/). It gives you all the capabilities and tools without you needing to understand how it all meshes and works in the background - just use the API and work on what is actually your business value.

**Reply by elirud** [reply]
Ye, sorry, it's poorly phrased. After looking around more I've decided that Kratos should definitely be a good first step. We have no intention of third party applications needing to login "with us" like with Google. If anything we would want to be able to use other sso providers in our service, which should be fine with Kratos as well. What we want it just a single service for people to login against with the same credentials, and that same service will take calls from other applications to then authorize incoming calls.

I got confused because I just went jumping right into this whole Oauth, OIDC rabbithole and wen't looking everywhere for answers. And some phrases might be used a bit differently between people, but ye, anyways. Hopefully this should work out well now when we got it set up. But after just focusing on the Ory ecosystem, and reading more into your different solutions, I feel like I have a good way forward. Please correct me if I'm wrong however and this actually isn't actually a good use case for Kratos.

**Reply by vinckr** [reply]
>What we want it just a single service for people to login against with the same credentials,

That is a great use case for Ory Kratos!

>that same service will take calls from other applications to then authorize incoming calls.

Can you elaborate on that a bit? Sounds more like a use case for [identity aware proxy](https://github.com/ory/oathkeeper) (basically takes incoming requests and authorizes against response from auth service e.g. Kratos).

**Reply by elirud** [reply]
Ye, so I'm definitely looking at Oathkeeper, but I think that it will have to wait a bit. But Oathkeeper and Keto is planned for getting added on top of Kratos now when we've got it set up. 

A question now when we have a basic version set up. We have most of our applications set up as serverless deployments under different subdomains. Which should be fine regarding the session cookie since I can set it for a shared TLD, but since I'm not finding anything about configuring the csrf cookie I'm assuming that the UI and Kratos will have to be on the same domain all together? So I'm thinking of having them both be deployed together and use nginx as a proxy for those two. Which should be good for adding authentication to the admin endpoints as well. But just wanted to check if I've missed something about the CSRF cookie?
